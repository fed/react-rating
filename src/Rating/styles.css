.Rating {
  display: inline-block;
}

.Rating__message {
  display: none;
}

.Rating__form {
  background-color: #fff;
  border: 1px solid #d2d0d2;
  border-radius: 5px;
  box-shadow: inset 0 0 4px 1px #c7c6c8;
  display: block;
  padding: 25px;
}

.Rating__title {
  color: #4a4a4a;
  font-family: "Helvetica Neue", sans-serif;
  font-size: 18px;
  font-weight: 600;
  margin-bottom: 16px;
}

.Rating__stars {
  border: 0;
  margin: 0 0 14px;
  padding: 0;
}

.Rating__button {
  background-color: #fff;
  border: 1px solid #d2d0d2;
  border-radius: 5px;
  color: #4a4a4a;
  cursor: pointer;
  font-family: "Helvetica Neue", sans-serif;
  font-size: 14px;
  padding: 4px 23px;
}

.Rating--confirmation .Rating__form {
  display: block;
  opacity: 0.4;
}

.Rating--confirmation .Rating__message {
  background-color: #7ed321;
  border-radius: 5px;
  color: #fff;
  display: block;
  font-family: "Helvetica Neue", sans-serif;
  font-size: 14px;
  font-weight: 600;
  margin-bottom: 6px;
  padding: 15px;
  text-align: center;
}
