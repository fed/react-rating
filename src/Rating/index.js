import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import uniqueId from 'lodash/uniqueId';
import './styles.css';

export default class Rating extends React.Component {
  constructor() {
    super();

    this.state = {
      // We generate a unique random ID so that we can have multiple
      // instances of the component on the same page.
      id: uniqueId('rating-'),
      rating: 0,
      submitted: false
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      rating: Number(event.target.value)
    });
  }

  handleSubmit(event) {
    event.preventDefault();

    this.props.onSubmit(this.state.rating);

    this.setState({
      submitted: true
    });
  }

  render() {
    const {id, rating, submitted} = this.state;
    const {title, buttonText, successMessage} = this.props;
    const className = classnames('Rating', {
      'Rating--confirmation': submitted
    });

    const stars = Array(5)
      .fill(0)
      .map((el, index) => {
        const value = index + 1;
        const name = `${id}-stars-${value}`;

        return (
          <span key={value} className="Rating__star">
            <input
              type="radio"
              name="stars"
              value={value}
              id={name}
              className="Rating__radio"
              onChange={this.handleChange}
              checked={rating === value} />

            <label
              className="Rating__label"
              htmlFor={name} />
          </span>
        );
      });

    return (
      <div className={className}>
        {
          submitted && (
            <div className="Rating__message">
              {successMessage}
            </div>
          )
        }

        <form className="Rating__form">
          <div className="Rating__title">{title}</div>
          <fieldset className="Rating__stars">{stars}</fieldset>
          <button type="submit" className="Rating__button" onClick={this.handleSubmit}>
            {buttonText}
          </button>
        </form>
      </div>
    );
  }
}

Rating.defaultProps = {
  title: 'Rate this product',
  buttonText: 'Apply',
  successMessage: 'Thanks for your rating!'
};

Rating.propTypes = {
  title: PropTypes.string,
  buttonText: PropTypes.string,
  successMessage: PropTypes.string,
  onSubmit: PropTypes.func.isRequired
}
