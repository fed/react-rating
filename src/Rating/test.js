import React from 'react';
import renderer from 'react-test-renderer';
import {mount, shallow} from 'enzyme';
import Rating from '.';
import noop from 'lodash/noop';

// We need to mock Lodash's uniqueId generator so that we can generate
// a valid snapshot and query the component.
jest.mock('lodash/uniqueId', () => jest.fn(() => 'rating-some-random-id'));

describe('Rating component', () => {
  it('renders without crashing', () => {
    mount(<Rating onSubmit={noop} />);
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <Rating onSubmit={noop} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('bootstraps with the correct initial state', () => {
    const wrapper = shallow(<Rating onSubmit={noop} />);

    expect(wrapper.state('rating')).toEqual(0);
    expect(wrapper.state('submitted')).toEqual(false);
  });

  it('renders the customised title, button text and success message if provided', () => {
    const wrapper = shallow(
      <Rating
        title="Rate us!"
        buttonText="Cast my vote"
        successMessage="Thanks for that"
        onSubmit={noop} />
    );
    const button = wrapper.find('.Rating__button');

    expect(wrapper.find('.Rating__title').text()).toEqual('Rate us!');
    expect(button.text()).toEqual('Cast my vote');

    // Now submit the form so that we can query the success message.
    button.simulate('click', { preventDefault() {} });
    expect(wrapper.find('.Rating__message').text()).toEqual('Thanks for that');
  });

  it('updates the internal state after selecting a rating', () => {
    const wrapper = mount(<Rating onSubmit={noop} />);
    const stars = wrapper.find('.Star__radio');
    const input = wrapper.find('.Rating__button');

    // First click on the first star and expect the rating to be 1.
    stars.first().simulate('change', { target: { value: 1 } });

    expect(wrapper.state('rating')).toEqual(1);
    expect(wrapper.state('submitted')).toEqual(false);

    // Then click on the fourth star (the one at index 3)
    // and expect the rating to be 4.
    stars.at(3).simulate('change', { target: { value: 4 } });

    expect(wrapper.state('rating')).toEqual(4);
    expect(wrapper.state('submitted')).toEqual(false);

    // Now submit the form and expect the correct rating value.
    input.simulate('click', { preventDefault() {} });

    expect(wrapper.state('rating')).toEqual(4);
    expect(wrapper.state('submitted')).toEqual(true);
  });

  it('calls the onSubmit function with the correct default value', () => {
    const spy = jest.fn();
    const wrapper = shallow(<Rating onSubmit={spy} />);
    const input = wrapper.find('.Rating__button');

    // Click submit without changing anything and
    // expect the onSubmit callback to be called with rating=0.
    input.simulate('click', { preventDefault() {} });

    expect(spy).toHaveBeenLastCalledWith(0);
  });

  it('calls the onSubmit function with the correct value', () => {
    const spy = jest.fn();
    const wrapper = mount(<Rating onSubmit={spy} />);
    const star = wrapper.find('#rating-some-random-id-stars-4');
    const input = wrapper.find('.Rating__button');

    // Now click on the fourth star and expect the callback
    // to be called with rating=4.
    star.simulate('change', { target: { value: 4 } });
    input.simulate('click', { preventDefault() {} });

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenLastCalledWith(4);
  });

  it('allows no further interaction once the form is submitted', () => {
    const wrapper = mount(<Rating onSubmit={noop} />);
    const stars = wrapper.find('.Star__radio');
    const input = wrapper.find('.Rating__button');

    // Select any random value and submit the form.
    // Then try to select any other rating. The internal state should not change.
    stars.at(2).simulate('change', { target: { value: 3 } });
    input.simulate('click', { preventDefault() {} });
    stars.at(4).simulate('change', { target: { value: 5 } });

    expect(wrapper.state('rating')).toEqual(3);
  });
});
