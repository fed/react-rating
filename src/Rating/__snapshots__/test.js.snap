// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating component renders correctly 1`] = `
<div
  className="Rating"
>
  <form
    className="Rating__form"
  >
    <div
      className="Rating__title"
    >
      Rate this product
    </div>
    <fieldset
      className="Rating__stars"
    >
      <span
        className="Star"
      >
        <input
          checked={false}
          className="Star__radio"
          disabled={false}
          id="rating-some-random-id-stars-1"
          name="stars-rating-some-random-id"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <label
          className="Star__label"
          htmlFor="rating-some-random-id-stars-1"
        />
      </span>
      <span
        className="Star"
      >
        <input
          checked={false}
          className="Star__radio"
          disabled={false}
          id="rating-some-random-id-stars-2"
          name="stars-rating-some-random-id"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <label
          className="Star__label"
          htmlFor="rating-some-random-id-stars-2"
        />
      </span>
      <span
        className="Star"
      >
        <input
          checked={false}
          className="Star__radio"
          disabled={false}
          id="rating-some-random-id-stars-3"
          name="stars-rating-some-random-id"
          onChange={[Function]}
          type="radio"
          value={3}
        />
        <label
          className="Star__label"
          htmlFor="rating-some-random-id-stars-3"
        />
      </span>
      <span
        className="Star"
      >
        <input
          checked={false}
          className="Star__radio"
          disabled={false}
          id="rating-some-random-id-stars-4"
          name="stars-rating-some-random-id"
          onChange={[Function]}
          type="radio"
          value={4}
        />
        <label
          className="Star__label"
          htmlFor="rating-some-random-id-stars-4"
        />
      </span>
      <span
        className="Star"
      >
        <input
          checked={false}
          className="Star__radio"
          disabled={false}
          id="rating-some-random-id-stars-5"
          name="stars-rating-some-random-id"
          onChange={[Function]}
          type="radio"
          value={5}
        />
        <label
          className="Star__label"
          htmlFor="rating-some-random-id-stars-5"
        />
      </span>
    </fieldset>
    <button
      className="Rating__button"
      disabled={false}
      onClick={[Function]}
    >
      Apply
    </button>
  </form>
</div>
`;
